File: dart/lib/src/packets/src/ai/im.dart
part of '../../packets.dart';

class ImPacket extends AIPacket {
  /// [timestamp] is the time of the packet.
  /// This is identified in the packet as `UNIX`
  final DateTime timestamp;

  /// [chatId] is the unique identifier for the chat.
  /// This is identified in the packet as `CHAT_ID`
  final String chatId;

  /// [message] is the chat message content.
  /// This is identified in the packet as `MESSAGE`
  ///
  /// Note, this message replaces any semicolons with the string `|||` to avoid packet parsing issues.
  final String message;

  /// [ImPacket] is the chat message packet
  ///
  /// This packet is part of the packet sent between Layrz services to stream AI conversations
  ImPacket({
    required this.timestamp,
    required this.chatId,
    required this.message,
  });

  /// [fromPacket] creates a [ImPacket] from a string packet in the format of `Layrz Protocol v3`.
  static ImPacket fromPacket(String raw) {
    if (!raw.startsWith('<Im>') || !raw.endsWith('</Im>')) {
      throw ParseException('Invalid identification packet, should be <Im>...</Im>');
    }

    final parts = raw.substring(4, raw.length - 5).split(';');
    if (parts.length != 4) {
      throw MalformedException('Invalid packet parts, should have 4 parts');
    }

    int? receivedCrc = int.tryParse(parts.last, radix: 16);
    int? calculatedCrc = calculateCrc("${parts.sublist(0, 3).join(';')};".codeUnits);

    if (receivedCrc != calculatedCrc) {
      throw CrcException('Invalid CRC, received: $receivedCrc, calculated: $calculatedCrc');
    }

    DateTime timestamp;
    try {
      timestamp = DateTime.fromMillisecondsSinceEpoch(int.parse(parts[0]) * 1000);
    } catch (e) {
      throw MalformedException('Invalid timestamp');
    }

    return ImPacket(
      timestamp: timestamp,
      chatId: parts[1],
      message: parts[2].replaceAll('|||', ';'),
    );
  }

  /// [toPacket] returns the packet in the format of `Layrz Protocol v3`.
  @override
  String toPacket() {
    String payload = '${(timestamp.millisecondsSinceEpoch / 1000).round()};';
    payload += '$chatId;';
    payload += '${message.replaceAll(';', '|||')};';
    String crc = calculateCrc(payload.codeUnits).toRadixString(16).padLeft(4, '0').toUpperCase();

    return '<Im>$payload$crc</Im>';
  }

  @override
  String toString() => toPacket();

  @override
  ImPacket copyWith({
    DateTime? timestamp,
    String? chatId,
    String? message,
  }) {
    return ImPacket(
      timestamp: timestamp ?? this.timestamp,
      chatId: chatId ?? this.chatId,
      message: message ?? this.message,
    );
  }
}

File: go/packets/ai/im.go
package ai

import (
	"errors"
	"fmt"
	"strconv"
	"strings"
	"time"

	"github.com/goldenm-software/layrz-protocol/go/v3/internal/wire"
)

// ImPacket is the AI message packet.
type ImPacket struct {
	// Timestamp of the packet
	Timestamp time.Time

	// ChatId is the unique chat identifier (UUID string)
	ChatId string

	// Message is the chat message content; semicolons are escaped as |||
	Message string
}

// FromPacket converts a raw <Im>...</Im> string to an ImPacket.
func (p *ImPacket) FromPacket(raw *string) error {
	if !strings.HasPrefix(*raw, "<Im>") || !strings.HasSuffix(*raw, "</Im>") {
		return errors.New("invalid packet, should be <Im>...</Im>")
	}

	body := (*raw)[4 : len(*raw)-5]

	rawCrc := body[len(body)-4:]
	body = body[:len(body)-4]

	receivedCrc, err := strconv.ParseUint(rawCrc, 16, 16)
	if err != nil {
		return errors.New("cannot convert CRC to integer")
	}

	calculatedCrc := wire.Calculate([]byte(body))
	if calculatedCrc != uint16(receivedCrc) {
		return fmt.Errorf("invalid CRC, received: %04X, calculated: %04X", receivedCrc, calculatedCrc)
	}

	body = strings.TrimSuffix(body, ";")
	parts := strings.Split(body, ";")
	if len(parts) != 3 {
		return errors.New("invalid packet, should have 3 parts")
	}

	rawTimestamp, err := strconv.ParseInt(parts[0], 10, 64)
	if err != nil {
		return errors.New("cannot convert timestamp to integer")
	}
	p.Timestamp = time.Unix(rawTimestamp, 0)
	p.ChatId = parts[1]
	p.Message = strings.ReplaceAll(parts[2], "|||", ";")

	return nil
}

// ToPacket converts an ImPacket to its wire representation.
func (p *ImPacket) ToPacket() *string {
	escapedMessage := strings.ReplaceAll(p.Message, ";", "|||")
	content := fmt.Sprintf("%d;%s;%s;", p.Timestamp.Unix(), p.ChatId, escapedMessage)
	crc := wire.Calculate([]byte(content))
	result := fmt.Sprintf("<Im>%s%04X</Im>", content, crc)
	return &result
}
